tests/Microsoft.Identity.Test.Integration.netfx/HeadlessTests/ConfidentialClientIntegrationTests.cs

```diff
@@ -649,6 +655,125 @@ private async Task RunOnBehalfOfTestAsync(LabResponse labResponse)
             Assert.IsNull(userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
         }
 
+        private async Task RunOnBehalfOfTestWithTokenCacheAsync(LabResponse labResponse)
+        {
+            LabUser user = labResponse.User;
+            string oboHost;
+            string secret;
+            string authority;
+            string publicClientID;
+            string confidentialClientID;
+            string[] oboScope;
+
+            oboHost = PublicCloudHost;
+            secret = _keyVault.GetSecret(TestConstants.MsalOBOKeyVaultUri).Value;
+            authority = TestConstants.AuthorityOrganizationsTenant;
+            publicClientID = PublicCloudPublicClientIDOBO;
+            confidentialClientID = PublicCloudConfidentialClientIDOBO;
+            oboScope = s_publicCloudOBOServiceScope;
+            
+            //TODO: acquire scenario specific client ids from the lab response
+
+            SecureString securePassword = new NetworkCredential("", user.GetOrFetchPassword()).SecurePassword;
+
+            var msalPublicClient = PublicClientApplicationBuilder.Create(publicClientID)
+                                                                 .WithAuthority(authority)
+                                                                 .WithRedirectUri(TestConstants.RedirectUri)
+                                                                 .WithTestLogging()
+                                                                 .Build();
+
+            var builder = msalPublicClient.AcquireTokenByUsernamePassword(oboScope, user.Upn, securePassword);
+
+            builder.WithAuthority(authority);
+
+            var authResult = await builder.ExecuteAsync().ConfigureAwait(false);
+
+            var confidentialApp = ConfidentialClientApplicationBuilder
+                .Create(confidentialClientID)
+                .WithAuthority(new Uri(oboHost + authResult.TenantId), true)
+                .WithClientSecret(secret)
+                .WithTestLogging()
+                .BuildConcrete();
+
+            var userCacheRecorder = confidentialApp.UserTokenCache.RecordAccess();
+
+            UserAssertion userAssertion = new UserAssertion(authResult.AccessToken);
+
+            string atHash = userAssertion.AssertionHash;
+
+            authResult = await confidentialApp.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            MsalAssert.AssertAuthResult(authResult, user);
+            Assert.AreEqual(atHash, userCacheRecorder.LastAfterAccessNotificationArgs.SuggestedCacheKey);
+
+            //Run OBO again. Should get token from cache
+            authResult = await confidentialApp.AcquireTokenOnBehalfOf(s_scopes, userAssertion)
+                .ExecuteAsync(CancellationToken.None)
+                .ConfigureAwait(false);
+
+            Assert.IsNotNull(authResult);
```

Comment:
A vbetter assertion would be to verify the `grant_type` that was used when talking to the /token endpoint. You can do that by injecting `HttpSnifferClientFactory` which will record all HTTP traffic. To retrieve the request payload, you'll have to add smth like: 

```csharp
                    req.Content.LoadIntoBufferAsync().GetAwaiter().GetResult();
                    var payload = req.Content.ReadAsStringAsync().GetAwaiter().GetResult(); 
 
   //  payload has grant_type

```